### ISSUE TITLE
Improvements to pretty printed reprs

### ISSUE THREAD

**Issue by jlstevens** [ioam/holoviews#149]
The repr of the following:

```python
import holoviews as hv
data = np.load('mandelbrot.npy')
(hv.Raster(data) * hv.HLine(y=200)).hist() + hv.Curve(data[200,:])
```
is currently:

```
:Layout
   .AdjointLayout.I :AdjointLayout   [AdjointLayout]
      :Histogram   [z]   (Frequency)
   .Curve.I         :Curve   [x]   (y)
```

There is no mention of the ``Raster`` or the ``HLine``. Removing the call to the hist method (so there is no ``AdjointLayout`` or ``Histogram``) restores the repr to the correct form.

**Comment by jlstevens** [comment]
In addition to pretty printing ``AdjointLayouts``, we need to make sure the ``OptionTree`` reprs match the ones shown in the SciPy paper.

**Comment by jlstevens** [comment]
I've closed the related issue #148 (``OptionTree`` reprs) and linked it back to this issue

**Comment by jlstevens** [comment]
I've worked on ``OptionTree`` quite a bit today due to this repr issue. It has been improved a fair bit because in the end we decided we want reprs of OptionTrees to work properly with ``eval``. Given that ``OptionTrees`` are core to the separation between data and presentation in HoloViews, this seemed fairly important to me.

Anyway, here is the repr we now need to update in the SciPy paper:

```python
OptionTree(groups=['plot', 'style', 'norm'],
   plot={
            'Curve ' : dict(aspect=1.25),
            'Layout' : dict(aspect_weight=1, sublabel_format='{Roman}')},

   style={
            'Curve' : dict(linewidth=3),
            'Image' : dict(cmap='Blues')}
)
```

Again, this doesn't use the pretty-printing system used everywhere else in HoloViews as this is a true repr (which is impossible/impractical for elements and container classes).

While talking about the SciPy paper, the way you get this reprs has also changed slightly - you now need to call the options to get the options specific to the current backend:

```python
options = hv.Store.custom_options()[content.id]
```

**Comment by jlstevens** [comment]
I've implemented the improvement mentioned in the edit above.

Personally, I don't see fixing the repr of ``AdjointLayout`` as sufficiently high priority to justify delaying 1.3.0.  I might get round to fixing this but if it is tricky, I am happy to leave this issue open for a while longer. What is a higher priority is updating the SciPy paper accordingly!

**Comment by jlstevens** [comment]
In commit 181ae4a I've implemented the following improvement: for the example above

```python
:Layout
   .AdjointLayout.I :AdjointLayout
      :Overlay
         .Raster.I :Raster   [x,y]   (z)
         .HLine.I  :HLine   [x,y]
      :Histogram   [z]   (Frequency)
   .Curve.I         :Curve   [x]   (y)
```

It isn't perfect but at least the ``Overlay``, ``Raster`` and ``HLine`` are no longer being hidden. I'm not really willing to spend more time on pretty printing ``AdjointLayouts`` given that we don't like this class much to begin with (see #177).

As the biggest issue (``__repr__`` of ``OptionTree``) has been addressed, I will now close this issue.
